numpy/_core/src/multiarray/arraytypes.c.src

```diff
@@ -1399,7 +1399,15 @@ HALF_to_@TYPE@(void *input, void *output, npy_intp n,
     @type@ *op = output;
 
     while (n--) {
-        *op++ = (@type@)npy_half_to_float(*ip++);
+        @type@ t;
+        if (npy_half_isnan(*ip)) {
+            t = (@type@)NPY_DATETIME_NAT;
+        }
+        else {
+            t = (@type@)npy_half_to_float(*ip++);
```

Comment:
Sorry, missed this:  This is actually incorrect.  It would be good to also modify the tests to end with a 0 value in the NaN array to catch this.

The `ip` pointer needs to be advanced no matter the branch.

Reply:
Oops, thanks! Sorry to miss that. I fixed this and modified the tests to catch this behaviour.

Reply:
Would you mind just moving the `ip++` into at the end together with the `*op++ = t`?  I don't think it is very nice to advance within the branches.

Reply:
I agree, just did that!